### PR TITLE
[UI Tests] For #116 #173: Verify the nav toolbar buttons

### PR DESCRIPTION
**Pull request by TejaswiKarasani**
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                verifyNavURLBarItems()
+
+                // Verify 'x' button actions
+                // verifyXButtonActions()
+
+                // Verify various items after returning back to the initial WebPage
+                // verifyNavURLBar()
+                verifyPageContent(defaultWebPage.content)
+            }.openHomeScreen {
+
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:61]: Nit: remove spacee

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    fun verifyTheNavToolBarButtonsTest() {
+        val defaultWebPage = TestAssetHelper.getGenericAsset(mockWebServer, 1)
+        val defaultWebPageTitle = "Test_Page_1"
+
+        // Visit a URL, wait until it's loaded
+        homeScreen {
+            navigationToolbar {
+            }.enterURLAndEnterToBrowser(defaultWebPage.url) {
+                verifyPageContent(defaultWebPage.content)
+
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:50]: Nit: remove space

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt (hunk 1 above)
Comment (TejaswiKarasani) [reply app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:61]: ok

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt (hunk 2 above)
Comment (TejaswiKarasani) [reply app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:50]: ok

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                // Verify the items of Nav URL Bar
+                verifyNavURLBarItems()
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:52]: I have thinking and there may be a different approach using the  Navigation toolbar robot. There you have defined already some of the  elements  you have defined here...
Also there is the 'x'  button. So you may want to add a [transition here](https://github.com/mozilla-mobile/fenix/blob/master/app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt#L44) to go back  to robot browser...
In summary, got to navigationToolbarRobot-> browserRobot -> navigationToolbarRobot -> browserRobot -> homescreenRobot
Hope that makes sense

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            navigationToolbar {
+            }.enterURLAndEnterToBrowser(defaultWebPage.url) {
+                verifyPageContent(defaultWebPage.content)
+                // Verify the items of Nav URL Bar
+                verifyNavURLBarItems()
+
+                // Verify 'x' button actions
+            }.openNavigationToolbar {
+            }.clearAddressBarActions {
+
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:56]: Nit: remove extra line

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                // Verify various items after returning back to the initial WebPage
+                verifyNavURLBar()
+                verifyPageContent(defaultWebPage.content)
+            }.openHomeScreen {
+                // Verify items on HomeScreen
+                verifyHomeScreen()
+                verifyExistingTabList()
+
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:64]: Nit: remove line

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt [hunk 6]
```diff
@@ -45,6 +46,24 @@ class NavigationToolbarRobot {
 
         val mDevice = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
 
+        fun clearAddressBarActions(interact: BrowserRobot.() -> Unit): BrowserRobot.Transition {
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt:49]: I would say  to rename this fun aas  clearAdddressBar... there are no more actions and it was  confusing for me first time looking at it. Just my opinion...

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt [hunk 7]
```diff
@@ -45,6 +46,24 @@ class NavigationToolbarRobot {
...
+            mDevice.waitNotNull(
+                Until.findObject(By.res("org.mozilla.fenix.debug:id/toolbar")),
+                waitingTime
+            )
+            urlBar().click()
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt:54]: Not sure why you  need this...once you are in browserRobot and tap on urlBar,  you are already on navigationToolBar... 🤔

Comment (TejaswiKarasani) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt:54]: To make the url bar into edit mode?

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt (hunk 6 above)
Comment (TejaswiKarasani) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt:49]: Yes, I will try to split it

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt (hunk 7 above)
Comment (TejaswiKarasani) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt:54]: Oh that was by mistake :/

Comment (isabelrios) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt:54]: not a mistake :) you can enter edit mode  there but it  is not necessary in this case.

Comment (TejaswiKarasani) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/NavigationToolbarRobot.kt:54]: Ok.

File: app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt [hunk 8]
```diff
@@ -370,6 +372,7 @@ class BrowserRobot {
         }
 
         fun openHomeScreen(interact: HomeScreenRobot.() -> Unit): HomeScreenRobot.Transition {
+            sessionLoadedIdlingResource = SessionLoadedIdlingResource()
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:375]: In addition to this, you missed this part: 
runWithIdleRes(sessionLoadedIdlingResource) {
                onView(
                        anyOf(
                                ViewMatchers.withResourceName("mozac_browser_toolbar_browser_actions")
                        )
                )
                        .check(matches(withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)))
            }

You should have to add that too...otherwise I guess it does not make any difference

Comment (TejaswiKarasani) [reply app/src/androidTest/java/org/mozilla/fenix/ui/robots/BrowserRobot.kt:375]: Okay

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            setDispatcher(AndroidAssetDispatcher())
+            start()
+        }
+    }
+
+    @After
+    fun tearDown() {
+        mockWebServer.shutdown()
+    }
+
```
Comment (isabelrios) [inline app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:43]: @TejaswiKarasani we should add here also a definition for the test, something similar as [here](https://github.com/mozilla-mobile/fenix/blob/master/app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt#L27) For example: Test to verify the navigation toolbar buttons in different screen after different actions

Comment (TejaswiKarasani) [reply app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:43]: @isabelrios , I mentioned it in the line 46 I think I should change that yeah you are right!

Comment (isabelrios) [reply app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:43]: oh sorry I did not see that description for the test, my bad. But better where it is now, thanks!

**Comment by isabelrios** [comment]
@AaronMT Modifications done, please if it looks good now, could you merge it? 
Tejaswi will be working on adding more tests to this test suite, so more changes could be done in a follow up if we something to modify now or later on.

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+import org.mozilla.fenix.helpers.AndroidAssetDispatcher
+import org.mozilla.fenix.helpers.HomeActivityTestRule
+import org.mozilla.fenix.helpers.TestAssetHelper
+import org.mozilla.fenix.ui.robots.homeScreen
+import org.mozilla.fenix.ui.robots.navigationToolbar
+
+/**
+ * This test suite will contain several tests covering different areas matching the smoketests defined in Test Rail
+ * https://testrail.stage.mozaws.net/index.php?/suites/view/3192&group_by=cases:section_id&group_order=asc.
+ */
```
Comment (AaronMT) [inline app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:23]: Let's setup the summary similar to outlying what is covered without having to click on the test-rail link
https://github.com/mozilla-mobile/fenix/blob/68017e5a00b09939d10b0f61e31d8d9e719dabbd/app/src/androidTest/java/org/mozilla/fenix/ui/StrictEnhancedTrackingProtectionTest.kt#L24

File: app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        }
+    }
+
+    @After
+    fun tearDown() {
+        mockWebServer.shutdown()
+    }
+
+    /**
+     *  Test goes through each basic functionality of NavigationToolbar of different screens after different actions
```
Comment (AaronMT) [inline app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:45]: The test title should be descriptive enough that the summary here is not really needed. ```verifyBasicNavigtionToolbarFunctionality``` is descriptive enough.

Comment (isabelrios) [reply app/src/androidTest/java/org/mozilla/fenix/ui/SmokeTest.kt:45]: Oh my bad, I suggested adding a description to the test too since we are going to have tests covering different areas, and to have more info. But if the title is enough, the comment can be removed
